relay-cardinality/src/redis.rs

```diff
@@ -0,0 +1,429 @@
+use std::{collections::HashSet, time::SystemTime};
+
+use itertools::Itertools;
+use relay_metrics::MetricNamespace;
+use relay_redis::{
+    redis::{self, ToRedisArgs},
+    Connection, RedisPool,
+};
+use relay_statsd::metric;
+
+use crate::{
+    limiter::{Entry, Limiter},
+    statsd::CardinalityLimiterCounters,
+    OrganizationId, Result,
+};
+
+/// Key prefix used for Redis keys.
+const REDIS_KEY_PREFIX: &str = "relay:cardinality";
+/// Maximum amount of arguments for a Redis `SADD` call.
+const REDIS_MAX_SADD_ARGS: usize = 1000;
+
+/// Scope for which cardinality will be tracked.
+///
+/// Currently this includes the organization and [namespace](`MetricNamespace`).
+/// Eventually we want to track cardinality also on more fine grained namespaces,
+/// e.g. on individual metric basis.
+#[derive(Clone, Copy, Debug, Hash, PartialEq, PartialOrd, Eq, Ord)]
+struct Scope {
+    /// The organization.
+    pub organization_id: OrganizationId,
+    /// The metric namespace, extracted from the metric bucket name.
+    pub namespace: MetricNamespace,
+}
+
+/// Implementation uses Redis sets to keep track of cardinality.
+pub struct RedisSetLimiter {
+    redis: RedisPool,
+    window: SlidingWindow,
+    #[cfg(test)]
+    time_offset: u64,
+}
+
+/// A Redis based limiter using Redis sets to track cardinality and membership.
+impl RedisSetLimiter {
+    /// Creates a new [`RedisSetLimiter`].
+    pub fn new(redis: RedisPool, window: SlidingWindow) -> Self {
+        Self {
+            redis,
+            window,
+            #[cfg(test)]
+            time_offset: 0,
+        }
+    }
+}
+
+impl RedisSetLimiter {
+    /// Checks the limits for a specific scope.
+    ///
+    /// Returns an iterator over all entries which have been accepted.
+    fn check_limits(
+        &self,
+        client: &mut RedisHelper,
+        scope: Scope,
+        mut entries: Vec<Entry>,
+        timestamp: u64,
+        limit: usize,
+    ) -> Result<impl Iterator<Item = Entry>> {
+        metric!(
+            counter(CardinalityLimiterCounters::RedisRead) += 1,
+            namespace = scope.namespace.as_str()
+        );
+        metric!(
+            counter(CardinalityLimiterCounters::RedisHashCheck) += entries.len() as i64,
+            namespace = scope.namespace.as_str()
+        );
+
+        let active_window = client
+            .read_set(&self.to_redis_key(&scope, self.window.active_time_bucket(timestamp)))?;
```

Comment:
As discussed: We could go straight with `SMISMEMBERS` instead of downloading the entire set.

Reply:
Unfortunately, we cannot do this, `SMISMEMBERS` was added in 6.2 and we still have instances on 5.X, another reason to look into the lua script